File: tests/src/test_session.cpp
#include <filesystem>
#include <fstream>
#include <memory>

#include "dataset.hpp"
#include "gtest/gtest.h"
#include "session.hpp"
#include "subject.hpp"

// Test fixture for Session class
class SessionTest : public ::testing::Test {
 protected:
  // Set up the test fixture
  void SetUp() override {
    // Create a temporary test directory
    this->test_dir = std::filesystem::temp_directory_path() / "dataset_test";
    std::filesystem::create_directory(test_dir);

    // Set up file paths
    participants_file = test_dir / "participants.tsv";
    participants_sidecar_file = test_dir / "participants.json";
    std::filesystem::create_directories(test_dir / "sub-01");
    std::filesystem::create_directories(test_dir / "sub-02");

    // Create participants table file
    std::ofstream participants_table(participants_file);
    participants_table << "participant_id\tname\n";
    participants_table << "sub-01\tJohn\n";
    participants_table << "sub-02\tAlice\n";
    participants_table.close();

    // Create participants sidecar file
    std::ofstream participants_sidecar(participants_sidecar_file);
    participants_sidecar << "{\n";
    participants_sidecar << "\t\"name\": {\n";
    participants_sidecar << "\t\t\"Description\": \"Name of participant\"\n";
    participants_sidecar << "\t}\n";
    participants_sidecar << "}\n";
    participants_sidecar.close();

    // Create the Dataset object)
    this->dataset = std::make_unique<Dataset>(this->test_dir, true);
    this->subject = this->dataset->get_subject(1);
  }

  // Tear down the test fixture
  void TearDown() override {
    // Remove the temporary test directory and files
    std::filesystem::remove_all(test_dir);
  }
  std::filesystem::path test_dir;
  std::filesystem::path participants_file;
  std::filesystem::path participants_sidecar_file;
  std::unique_ptr<Dataset> dataset;
  std::unique_ptr<Subject> subject;
};

// Test the path() method of Session
TEST_F(SessionTest, PathTest) {
  // Create a Session object with index 1
  Session session(*subject, 1);

  // Expected path: subject path + session id
  std::filesystem::path expectedPath = subject->path() / "ses-01";

  // Check if the calculated path matches the expected path
  EXPECT_EQ(session.path(), expectedPath);
}

// Test the prefix() method of Session
TEST_F(SessionTest, PrefixTest) {
  // Create a Session object with index 2
  Session session(*subject, 2);

  // Expected prefix: subject id + "_" + session id
  std::string expectedPrefix = subject->id() + "_" + session.id();

  // Check if the calculated prefix matches the expected prefix
  EXPECT_EQ(session.prefix(), expectedPrefix);
}

File: tests/test_session.py
import pytest
import shutil
import tempfile

from pathlib import Path
from libbids import Dataset
from libbids.clibbids import Session, Subject  # type: ignore
from typing import cast


# Test fixture for Session class
class TestSession:
    # Set up the test fixture
    @pytest.fixture(autouse=True)
    def setup(self):
        self.test_dir = Path(tempfile.mkdtemp())
        self.participants_file = self.test_dir / "participants.tsv"
        self.participants_sidecar_file = self.test_dir / "participants.json"
        (self.test_dir / "sub-01").mkdir()
        (self.test_dir / "sub-02").mkdir()

        # Create participants table file
        with open(self.participants_file, "w") as file:
            file.write("participant_id\tname\n")
            file.write("sub-01\tJohn\n")
            file.write("sub-02\tAlice\n")

        # Create participants sidecar file
        with open(self.participants_sidecar_file, "w") as file:
            file.write("{\n")
            file.write('\t"name": {\n')
            file.write('\t\t"Description": "Name of participant"\n')
            file.write("\t}\n")
            file.write("}\n")

        self.dataset = Dataset(self.test_dir, True)
        self.subject = self.dataset.get_subject(1)

        yield

        # Clean up the temporary test directory and files
        shutil.rmtree(self.test_dir)

    # Test the path() method of Session
    def test_path_test(self) -> None:
        # Create a Session object with index 1
        session: Session = Session(self.subject, 1)

        # Expected path: subject path + session id
        expected_path: Path = cast(Subject, self.subject).path / "ses-01"

        # Check if the calculated path matches the expected path
        assert session.path == expected_path

    # Test the prefix() method of Session
    def test_prefix(self) -> None:
        # Create a Session object with index 2
        session: Session = Session(self.subject, 2)

        # Expected prefix: subject id + "_" + session id
        expected_prefix: str = cast(Subject, self.subject).id + "_" + session.id

        # Check if the calculated prefix matches the expected prefix
        assert session.prefix == expected_prefix

    # test the label
    def test_label(self) -> None:
        # Create a Session object with index 3
        session: Session = Session(self.subject, 3)

        # Check if the calculated prefix matches the expected prefix
        assert session.label == "03"

    # test padding
    def test_padding(self) -> None:
        # Create a Session object with index 4
        session: Session = Session(self.subject, 4)

        # Check if the calculated prefix matches the expected prefix
        assert session.padding == 2

    # test index
    def test_index(self) -> None:
        # Creat a Session object with index 5
        session: Session = Session(self.subject, 5)

        # Check if the calculated index matches
        assert session.index == 5
